### ISSUE TITLE
Get current test name

### ISSUE THREAD

**Issue by majkinetor** [pester/Pester#1611]
Is it possible to get current test name (the one available as  `ExpandedName` after test runs) during test exeuction ?

I want to create generic helper that will save some debuging information in the file that is named by the test name.

**Comment by JustinGrote** [comment]
@nohwnd there is a TestName variable present in the It context but it isn't being populated. Bug maybe? No reason it shouldn't have all the information it needs to populate it prior to It execution
![image](https://user-images.githubusercontent.com/15258962/85336306-417b1900-b493-11ea-8588-ad859e13294a.png)

**Comment by efie45** [comment]
If this is the new 5.0.2 version:
The code that executes directly inside of the Describe block is evaluated during the new 'Discovery' phase and isn't available when the tests run. Try throwing your script sourcing inside of a BeforeAll or BeforeEach block right inside of the Describe block and before your It block.

**Comment by mkarpuk** [comment]
So, is there a way to get current test name at some place (BeforeEach or inside It block)?
I didn't get this from the thread, sorry. Would like to send this information to the ElasticSearch for tracing purposes.

**Comment by majkinetor** [comment]
Yeah, for Elastic you would definitelly need to obtain it, that is what I needed it for among other things. The other case was sending results data to InfluxDb for metric purposes and flaky test analysis across runs (also for timings) and for that it was enough to send everything after the run using results object. 

I hardcoded test names for now on few places that I needed it the most, but this is something required for systemic solution (and also supported by some other test frameworks).

**Comment by JustinGrote** [comment]
@mkarpuk as far as I can tell you can't at the moment and I'm pretty sure its a bug, you could workaround it by just setting a variable within each test e.g. $SCRIPT:CurrentTestName = 'mytest1' for now.

**Comment by majkinetor** [comment]
> $SCRIPT:CurrentTestName = 'mytest1'

That will get ugly pretty fast. Maybe a preprocessor is a better idea - it can inject such things before execution.

**Comment by JustinGrote** [comment]
@majkinetor like I said it's a temporary workaround not a prescriptive guidance, there is a testname variable that should be populated. If I find some time maybe I'll put together a PR.

**Comment by nohwnd** [comment]
You could do this, if you are not afraid to use internal apis, that might change in the future

```powershell
& (get-module pester) {

    $state.Plugin += (New-PluginObject -Name "Logger" -EachTestSetupStart { 
        $Test = $Context.Test 

        Write-Host "name: $($Test.Name), path: $($Test.Path)"
    } )
}

Describe "a" { 

    It "b" { 
        
    }

    It "c" {

    }
}
```


```shell

Starting discovery in 1 files.
Discovery finished in 40ms.
name: b, path: a b
name: c, path: a c
[+] C:\Users\jajares\Desktop\fufuffufuf.tests.ps1 261ms (55ms|178ms)
Tests completed in 264ms
Tests Passed: 2, Failed: 0, Skipped: 0 NotRun: 0
```

**Comment by majkinetor** [comment]
Any chance we can have something normal ? :) Something like `$TestName` variable within test or similar ...

Interesting syntax - pretty sure I have never seen it in last decade on PowerShell :)

**Comment by JustinGrote** [comment]
@majkinetor the syntax is how you use the call operator to run a comand within the module scope, it's a neat trick especially when testing modules.

Here is an example that is more what you are looking for, WARNING it's a kludge

```powershell
function Save-TestNameToVariable {
    & (get-module pester) {

        $state.Plugin += (New-PluginObject -Name "Logger" -EachTestSetupStart { 
            $GLOBAL:TestName = $Context.Test -replace '^\[ERR\] '
            #Set-Variable -Scope Global -Name $TestName -Value ($Context.Test -replace '^\[ERR\] ')
        } )
    }
}

Describe "a" { 
    Save-TestNameToVariable

    It "b" { 
        $TestName | Should -Be 'b'
    }

    It "c" {
        $TestName | Should -Be 'c'
    }
}
```

![image](https://user-images.githubusercontent.com/15258962/87064302-5f02ef00-c1c4-11ea-8f78-d96741c14726.png)

**Comment by majkinetor** [comment]
Yeah, thats more nice. I am not sure if it will work will paralel execution (https://github.com/pester/Pester/issues/1270)

**Comment by JustinGrote** [comment]
@majkinetor yeah it's not thread safe, you might hit a race condition unless the plugin runs within the runspace of the test, I'm not sure when Parallel "kicks" in in terms of the plugins.

**Comment by simonsabin** [comment]
How fragile/hacky is doing this ` $____Pester.CurrentTest.Name`

**Comment by nohwnd** [comment]
@simonsabin that's recent addition. I plan to embrace it, but did not settle on the name entirely. Probably will rename it to just Pester later. Other than that, if you don't modify the state, and take the risk of the variable being renamed later you should be good to go. I think you can assume that you will have access to the test name in one way or another in the future.
